### DISCUSSION TITLE
#[ORM\HasLifecycleCallbacks] and EntityListner for the same entity

### DISCUSSION THREAD

**Discussion by armelo24** [symfony/symfony#52462] [Q&A] [answered]
Hello, I have an entity "Category" on which I define the creation date "createdAt" automatically and only during persistence via the "prePersist" event of the entity's life cycle with #[ORM\HasLifecycleCallbacks] . Everything works fine for that.
Then I needed to define for the same entity an "EntityListner" which would allow me to automatically calculate the slug of the category before its persistence or its update: "prePersist" and "preUpdate".
But this listner is simply ignored. I do not know why.
can you help me ?

**Comment by themark147** [comment]
Hey

U didnt forgot to define listener as listener via AsEntityListener or AsDoctrineListener attribute ? :thinking: https://symfony.com/doc/current/doctrine/events.html#doctrine-lifecycle-listeners

**Reply by armelo24** [reply]
Thx for your replie. I all ready did that it didn't work.
```php
<?php

namespace App\Service\Slugger\EntityListner;

use App\Entity\Category;
use Doctrine\Bundle\DoctrineBundle\Attribute\AsEntityListener;
use Doctrine\ORM\Events;
use Doctrine\Persistence\Event\LifecycleEventArgs;
use Symfony\Component\String\Slugger\SluggerInterface;

#[AsEntityListener(event: Events::prePersist, entity: Category::class)]
#[AsEntityListener(event: Events::preUpdate, entity: Category::class)]
class SlugComputerEntityListener
{
    public function __construct(private SluggerInterface $slugger) {}

    public function prePersist(Category $category, LifecycleEventArgs $event): void
    {
        $category->setCreatedAt();
        $category->computeSlug($this->slugger);
    }

    public function preUpdate(Category $category, LifecycleEventArgs $event): void
    {
        $category->computeSlug($this->slugger);
    }
}
```

**Reply by themark147** [reply]
u tried defined methods in attributes as well?

```
#[AsEntityListener(event: Events::prePersist, method: 'prePersist', entity: Category::class)]
#[AsEntityListener(event: Events::preUpdate, method: 'preUpdate', entity: Category::class)]
```

**Comment by armelo24** [comment]
Yes i did it but the same error. I also change LifeCycleEventArgs to prePersistEventArgs and PreUpdateEventArg in each method but nothing change. the same error bellow.
![image](https://github.com/symfony/symfony/assets/42834337/ec891456-d22e-4430-9b73-29fd0aff0b87)

`<?php

namespace App\Service\Slugger\EntityListner;

use App\Entity\Category;
use Doctrine\Bundle\DoctrineBundle\Attribute\AsEntityListener;
use Doctrine\ORM\Event\PrePersistEventArgs;
use Doctrine\ORM\Event\PreUpdateEventArgs;
use Doctrine\ORM\Events;
use Doctrine\Persistence\Event\LifecycleEventArgs;
use Symfony\Component\String\Slugger\SluggerInterface;

#[AsEntityListener(event: Events::prePersist, method: 'prePersist', entity: Category::class)]
#[AsEntityListener(event: Events::preUpdate, method: 'preUpdate', entity: Category::class)]
class SlugComputerEntityListener
{
    public function __construct(private SluggerInterface $slugger) {}

    public function prePersist(Category $category, PrePersistEventArgs $event): void
    {
        $category->setCreatedAt();
        $category->computeSlug($this->slugger);
    }

    public function preUpdate(Category $category, PreUpdateEventArgs $event): void
    {
        $category->computeSlug($this->slugger);
    }
}
`

`<?php

namespace App\Entity;

use App\Entity\Trait\CreatedAtTrait;
use App\Entity\Trait\SlugTrait;
use App\Repository\CategoryRepository;
use App\Service\Slugger\EntityListner\SlugComputerEntityListener;
use App\Service\Slugger\SlugComputerInterface;
use Doctrine\Common\Collections\ArrayCollection;
use Doctrine\Common\Collections\Collection;
use Doctrine\ORM\Mapping as ORM;
use Doctrine\ORM\Mapping\EntityListeners;
use Symfony\Bridge\Doctrine\Validator\Constraints\UniqueEntity;
use Symfony\Component\String\Slugger\SluggerInterface;

// #[EntityListeners([SlugComputerEntityListener::class])]
#[UniqueEntity('slug')]
// #[ORM\HasLifecycleCallbacks]
#[ORM\Entity(repositoryClass: CategoryRepository::class)]
class Category implements SlugComputerInterface
{
    use CreatedAtTrait;
    use SlugTrait;

    #[ORM\Id]
    #[ORM\GeneratedValue]
    #[ORM\Column]
    private ?int $id = null;

. . . .
}
`

**Comment by armelo24** [comment, marked as answer]
OK, I've solved my problem. An incongruence between the namespace of my EntityListener and the folder it was in.  To detect this, I opted to register the Listener directly in the "service.yaml" config file instead of the php attributes and when I refreshed the page an error appeared saying that the service container couldn't load the file because it didn't exist. 

`namespace App\Service\Slugger\EntityListener;`
instead of 
`namespace App\Service\Slugger\EntityListner;`
